extension/goplugin/environment.go

```diff
@@ -482,26 +479,73 @@ func handleEventForEnv(env IEnvironment, event string, requestContext map[string
 	return nil
 }
 
-func hasCancel(requestContext map[string]interface{}) bool {
+func hasInterrupt(requestContext map[string]interface{}) bool {
 	_, cancelFound := requestContext["context"]
 	return cancelFound
 }
 
-func addCancel(env IEnvironment, event string, requestContext map[string]interface{}, done <-chan bool) {
-	ctx, cancel := buildCancel(env, event)
-	cancelOnPeerDisconnect(requestContext, cancel, done)
-	requestContext["context"] = ctx
+// interrupt decorates a request identified by requestContext with a context.Context
+// inserted into requestContext["context"]. This context will be canceled when
+// a. extension timeout expires
+// b. HTTP client disconnects
+// Interrupts are constructed
+// Users of interrupt are supposed to call Cleanup() when request processing is done
+type interrupt struct {
+	env            IEnvironment
+	event          string
+	requestContext map[string]interface{}
+	done           chan struct{}
+	ctx            context.Context
+	cancel         context.CancelFunc
+}
+
+func newInterrupt(env IEnvironment, event string, requestContext map[string]interface{}) *interrupt {
+	ctx, cancel := context.WithCancel(context.Background())
+	interrupt := &interrupt{env, event, requestContext, make(chan struct{}, 1), ctx, cancel}
```

Comment:
use named field initialization

Reply:
as discussed offline: only the `done` channel field lacks a proper name, fixed